src/MockCollection.php

```diff
@@ -80,24 +87,60 @@ public function updateOne($filter, $update, array $options = [])
         $matcher = $this->matcherFromQuery($filter);
         foreach ($this->documents as $i => &$doc) {
             if ($matcher($doc)) {
-                foreach ($update['$set'] ?? [] as $k => $v) {
-                    $doc[$k] = $v;
-                }
+                $this->updateCore($doc,$update);
                 return;
             }
         }
+        $this->updateUpsert($filter,$update,$options,false);
     }
 
     public function updateMany($filter, $update, array $options = [])
     {
         $matcher = $this->matcherFromQuery($filter);
+        $anyUpdates = false;
         foreach ($this->documents as $i => &$doc) {
             if (!$matcher($doc)) {
                 continue;
             }
 
-            foreach ($update['$set'] ?? [] as $k => $v) {
-                $doc[$k] = $v;
+            $this->updateCore($doc,$update); 
+            $anyUpdates = $anyUpdates || true;
+        }
+
+        $this->updateUpsert($filter,$update,$options,$anyUpdates);
+        return $anyUpdates;
+    }
+
+    private function updateUpsert($filter, $update, $options, $anyUpdates) {
+        if (array_key_exists('upsert', $options)) {
+            if ($options['upsert'] && !$anyUpdates) {
+                if (array_key_exists('$set',$update)) {
+                    $documents = [array_merge($filter, $update['$set'])];
+                } else {
+                    $documents = [$update];
+                }
+                $this->insertMany($documents, $options);
+            }
+        }
+    }
+
+    private function updateCore(&$doc, $update)
+    {
+        // The update operators are required, as exemplified here:
+        // http://mongodb.github.io/mongo-php-library/tutorial/crud/
+        $supported = [ '$set', '$unset' ];
+        $unsupported = array_diff(array_keys($update),$supported);
+        if(count($unsupported)>0) throw new \Exception("Unsupported update operators found: ".implode(', ',$unsupported));
+
+        foreach ($update['$set'] ?? [] as $k => $v) {
+            $doc[$k] = $v;
+        }
+
+        if (array_key_exists('$unset',$update)) {
```

Comment:
Technically, the `array_key_exists` check is superflous; the `??` operator in the nextline also contains an implicit `isset` check, so this would work one way or the other.